### PR TITLE
extension and further discussion, second try

### PR DESCRIPTION
**Pull request by gluk256**
this file was missing from my previous PR

### CODE REVIEW

File: extension.md [hunk 1]
```diff
@@ -0,0 +1,28 @@
+## Further Analysis
+
+### Economics of Cross-Witness
+
+If any participant cheats, his money at stake must be enough to cover for all the expenses. The contract denomination must be grater than aggregated transaction costs of all the participants, plus potential arbitration costs. The aggregated costs are proportionate to the number of participants, and therefore there must be a limit to the number of participants for each denomination of smart contract. If the number of people exceeds the contract limit, it might result in shortages/queues. This gives the malicious entities an opportunity to launch DDoS attack. One of the possible solutions: multiple contracts + one "router" contract, through which all the TX must be initiated.
```
Comment (gavofyork) [inline extension.md]: grater -> greater

File: extension.md [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+
+### Random Numbers Generation (RNG)
+
+How can we prove, that router contract is honest ("nothing up my sleeve"), and besides, may not be manipulated by miners or anyone else?
+After the contract is activated, everybody must sign the next block's hash with his public key, and then publish the signature. Those signatures, if valid, will be in turn concatenated or XORed, and the result hashed, thus producing the seed for RNG.
+
+BTW, this algorithm might be also used in gambling dapps. They might even call it "rolling the dice" for marketing purposes, as it gives the participants the same illusion as physically rolling the dice -- illusion of some kind of personal control over the RNG.
```
Comment (gavofyork) [inline extension.md]: this has already been invented - search for random DAO or "randao".

File: extension.md [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+
+### Onion Concept
+
+We assume, that multiple iterations will be necessary to acheive the desired probability of untracebility. If we could do several iterations at once, we might save on transaction costs, execution time, and -- which might be even more important -- mental costs of initiating transaction, choosing the available contract in case of queues, and managing the multiple accounts (although this could be automated, but still requires external overhead).
+
+Here is one of the possible solutions. During the Witness Selection step, everybody should select certain (predefined) number of witnesses, then repeatedly encrypt his data with their public keys, then publish the resulting encrypted data ("onion"). Next step remains the same, except that it must be repeated as many times, as required to "unwind the onion".
```
Comment (gavofyork) [inline extension.md:11]: yeah - i considered this myself. the multiple rounds may end up being too onerous for most uses though; rounds are expensive as they require all participants to SSTORE and would likely take a safe minimum of 2 minutes (transaction propagation and block inclusion often takes > 4 blocks).

File: extension.md [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+
+### Onion Economics
+
+If the original version of Cross-Witness algorithm runs smoothly, it requires the following transactions: 
+- initial money transfer with attached public key
+- publication of encrypted dest
+- publication of decrypted dest
+- signing and dissolution
+
+Now, if we add one extra iteration (in case of two-layer onion), we only add one more TX -- publication of decrypted onion layer. If we add 4 iterations (five-layer onion), the TX costs will only double, but probability of failure will be p^5. However, arbitration costs increase dramatically. So, the economic calculations should be corrected correspondingly. It might also mean, that onion extension will be economically viable only for large denomination contracts.
```
Comment (gavofyork) [inline extension.md:21]: Indeed :)

**Comment by gluk256** [comment]
Yeah, randao is indeed already invented -- thanks for pointing out :)
I have left the reference to randao, though, so that explaination is coherent.

**Pull request closed without merging**